### ISSUE TITLE
kubenet: repeated ebtables rule causes bad network performances (v1.16 and v1.17 only)

### ISSUE THREAD

**Issue by alban** [kubernetes/kubernetes#89633]
kubenet: repeated ebtables rule causes bad network performances (v1.16 and v1.17 only)

**What happened**:

Updating Kubernetes from 1.14 to 1.16 caused applications to have bad network performances. Other parameters were not changed: same distribution, same kernel, etc.

**What you expected to happen**:

Applications' network performances should not drop when updating Kubernetes.

**How to reproduce it (as minimally and precisely as possible)**:

Run applications with the kubenet network plugin.

**Anything else we need to know?**:

The [following line](https://github.com/kubernetes/kubernetes/blob/v1.16.4/pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go#L275) is added in v1.16.0 by https://github.com/kubernetes/kubernetes/commit/dba434c4ba0bf39275f25b3833fcc5f6eefa9379 (/cc @khenidak) and was removed for v1.18.0 via https://github.com/kubernetes/kubernetes/commit/c5002a348e3021861730502011197614477adb39:

```
cidr.IP[len(cidr.IP)-1] += 1 // Set bridge address to first address in IPNet
```

This is bad because `cidr.String()` will print a CIDR like `10.2.207.1/24`: the `.1` and the `/24` are contradicting each other. See demonstration in this small Go program:

<details>

```
package main

import (
        "fmt"
        "log"
        "net"
)

func main() {
        _, cidr, err := net.ParseCIDR("10.2.207.0/24")
        if err != nil {
                log.Fatal(err)
        }
        fmt.Println(cidr) // Prints "10.2.207.0/24"
        cidr.IP[len(cidr.IP)-1] += 1 // Set bridge address to first address in IPNet
        fmt.Printf("%s\n", cidr.String()) // Prints "10.2.207.1/24"
}
```

</details>

This CIDR is used for ebtables rules in the `KUBE-DEDUP` chain. Fortunately, the ebtables CLI understands CIDRs like `10.2.207.1/24` and silently converts it to `10.2.207.0/24`. However, this breaks [checkIfRuleExists()](https://github.com/kubernetes/kubernetes/blob/v1.16.4/pkg/util/ebtables/ebtables.go#L163) because the rule printed by `ebtables -L` has a different format. See the WARNING in the code:
```
// checkIfRuleExists takes the output of ebtables list chain and checks if the input rules exists
// WARNING: checkIfRuleExists expects the input args matches the format and sequence of ebtables list output
func checkIfRuleExists(listChainOutput string, args ...string) bool {
```

This causes the rule to be repeated. When the cluster runs for a long time, the rule can be repeated thousands of times. This causes the rule to be evaluated thousands of times for each network packet, causing high CPU usage and low network performance.

This bug was identified using the [BCC CPU profiler](https://github.com/iovisor/bcc/blob/master/tools/profile_example.txt) in the following way:

<details>

Run the following command:

```
docker run --rm -ti --privileged --net=host --pid=host \
    -v /usr/src:/usr/src \
    -v /lib/modules:/lib/modules \
    docker.io/kinvolk/bcc \
    /usr/share/bcc/tools/profile -d -f -K 10
```

Among other things, the following stack was often printed:
```
thread.rb:70;xen_hvm_callback_vector;xen_evtchn_do_upcall;irq_exit;__softirqentry_text_start;net_rx_action;ena_io_poll;ena_clean_rx_irq;napi_gro_receive;netif_receive_skb_internal;__netif_receive_skb;__netif_receive_skb_core;ip_rcv;ip_rcv_finish;ip_forward;ip_forward_finish;ip_output;ip_finish_output;ip_finish_output2;dev_queue_xmit;__dev_queue_xmit;__qdisc_run;sch_direct_xmit;dev_hard_start_xmit;br_dev_xmit;br_forward;__br_forward;nf_hook_slow;ebt_out_hook;ebt_in_hook;ebt_ip_mt;ebt_ip_mt 3
```
This hinted that the bad performances were around the kernel function [ebt_ip_mt()](https://github.com/torvalds/linux/blob/master/net/bridge/netfilter/ebt_ip.c). This function is related to the bridging firewall ebtables, so I looked at the ebtables rules and noticed thousands of lines repeating the same rule in the `KUBE-DEDUP` chain.

</details>

**Environment**:
- Kubernetes version (use `kubectl version`): 1.16.4
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Network plugin and version (if this is a network-related bug): kubenet
- Others: N/A

/cc @iaguis

**Comment by tedyu** [comment]
```
in this small Go program
```
Please attach the program.
```
using the BCC CPU profiler in the following way:
```
I don't see description for troubleshooting procedure.

**Comment by alban** [comment]
@tedyu I included them in the collapsible section `<details> ... </details>`. But I can add them in full here as well.

### Go program

```
package main

import (
        "fmt"
        "log"
        "net"
)

func main() {
        _, cidr, err := net.ParseCIDR("10.2.207.0/24")
        if err != nil {
                log.Fatal(err)
        }
        fmt.Println(cidr) // Prints "10.2.207.0/24"
        cidr.IP[len(cidr.IP)-1] += 1 // Set bridge address to first address in IPNet
        fmt.Printf("%s\n", cidr.String()) // Prints "10.2.207.1/24"
}
```

### Description for troubleshooting procedure

Run the following command:

```
docker run --rm -ti --privileged --net=host --pid=host \
    -v /usr/src:/usr/src \
    -v /lib/modules:/lib/modules \
    docker.io/kinvolk/bcc \
    /usr/share/bcc/tools/profile -d -f -K 10
```

Among other things, the following stack was often printed:
```
thread.rb:70;xen_hvm_callback_vector;xen_evtchn_do_upcall;irq_exit;__softirqentry_text_start;net_rx_action;ena_io_poll;ena_clean_rx_irq;napi_gro_receive;netif_receive_skb_internal;__netif_receive_skb;__netif_receive_skb_core;ip_rcv;ip_rcv_finish;ip_forward;ip_forward_finish;ip_output;ip_finish_output;ip_finish_output2;dev_queue_xmit;__dev_queue_xmit;__qdisc_run;sch_direct_xmit;dev_hard_start_xmit;br_dev_xmit;br_forward;__br_forward;nf_hook_slow;ebt_out_hook;ebt_in_hook;ebt_ip_mt;ebt_ip_mt 3
```
This hinted that the bad performances were around the kernel function [ebt_ip_mt()](https://github.com/torvalds/linux/blob/master/net/bridge/netfilter/ebt_ip.c). This function is related to the bridging firewall ebtables, so I looked at the ebtables rules and noticed thousands of lines repeating the same rule in the `KUBE-DEDUP` chain.

**Comment by tedyu** [comment]
thanks for the information

**Comment by alban** [comment]
Additional troubleshooting information:

In order to see which ebtables command is executed by the kubenet plugin, I run the [BCC execsnoop tool](https://github.com/iovisor/bcc/blob/master/tools/execsnoop_example.txt) running for a while:

```
docker run --rm -ti --privileged --net=host --pid=host -v /usr/src:/usr/src -v /lib/modules:/lib/modules -v /sys/kernel/debug:/sys/kernel/debug docker.io/kinvolk/bcc /usr/share/bcc/tools/execsnoop  | grep -E 'COMM|ebtables'
```

And then I noticed the following:
```
ebtables         5260   2411     0 /sbin/ebtables --version
ebtables         5261   2411     0 /sbin/ebtables -t filter -L KUBE-DEDUP
ebtables         5262   2411     0 /sbin/ebtables -t filter -L OUTPUT --Lmac2
ebtables         5263   2411     0 /sbin/ebtables -t filter -L KUBE-DEDUP --Lmac2
ebtables         5264   2411     0 /sbin/ebtables -t filter -L KUBE-DEDUP --Lmac2
ebtables         5265   2411     0 /sbin/ebtables -t filter -A KUBE-DEDUP -p IPv4 -s xx:xx:xx:xx:xx:xx -o veth+ --ip-src 10.2.207.1/24 -j DROP
```

I've edited the mac address. We can see ebtables executed with the parameter `10.2.207.1/24`.

**Comment by ipochi** [comment]
The mentioned fix [c5002a3](https://github.com/kubernetes/kubernetes/commit/c5002a348e3021861730502011197614477adb39) has been back ported to 1.17.3 ([4a8b94e](https://github.com/kubernetes/kubernetes/commit/4a8b94ed2b10086b11fde5522fb50c121ffa178a)) and 1.16.8 ([32a8be1](https://github.com/kubernetes/kubernetes/commit/32a8be11426a61e61af508f700a615028f609ec9)) releases.

**Comment by alban** [comment]
Thanks @ipochi! I will close the issue then.
